Source/WebKit/UIProcess/WebProcessPool.cpp

```diff
@@ -1223,8 +1223,12 @@ Ref<WebPageProxy> WebProcessPool::createWebPage(PageClient& pageClient, Ref<API:
     RefPtr<WebProcessProxy> process;
     auto lockdownMode = pageConfiguration->lockdownModeEnabled() ? WebProcessProxy::LockdownMode::Enabled : WebProcessProxy::LockdownMode::Disabled;
     RefPtr relatedPage = pageConfiguration->relatedPage();
-    // FIXME: If site isolation is enabled, this needs to get the correct process instead of assuming we always want the process of the related page.
-    if (relatedPage && !relatedPage->isClosed() && relatedPage->hasSameGPUAndNetworkProcessPreferencesAs(pageConfiguration)) {
+
+    auto openerFrameID = pageConfiguration->openerFrameID();
+    if (pageConfiguration->preferences().siteIsolationEnabled() && openerFrameID) {
+        if (RefPtr openerFrame = WebFrameProxy::webFrame(*openerFrameID))
+            process = &openerFrame->process();
```

Comment:
`process` is not set when `!openerFrame`, which will fail the assertion `ASSERT(process);` below.

Reply:
I think that is ok. `openerFrame` should always exist if `openerFrameID` does.

Reply:
I don't think that's correct assumption because this function is invoked with configuration object passed by client, and client can technically set `openerFrameID` to anything. 
Even client uses the same configuration object created by WebKit, they can create the WebView asynchronously, so the frame / page may not exist when client invokes the function (see `UIDelegate::UIClient::createNewPage`).

If we are sure `openerFrame` exists, this should be written without the if statement for clarity:
```
RefPtr openerFrame = WebFrameProxy::webFrame(*openerFrameID)
RELEASE_ASSERT(openerFrame);
process = &openerFrame->process();
```

Reply:
I see. Then yeah that assumption is wrong.

Reply:
decidePolicyForNavigationAction is async, so the opener could actually be destroyed before this, so we don't want to RELEASE_ASSERT.  I'll make it fall back to the other process selection code if that happens.

Reply:
The frame may also have navigated to a different origin.  I'm going to remember the process instead of the frame.